src/NATS.Client.Core/INatsSerializer.cs

```diff
@@ -611,54 +622,71 @@ public void Serialize<T>(IBufferWriter<byte> bufferWriter, T? value)
             throw new NatsException($"Can't deserialize {typeof(T)}");
         }
 
-        return _next.Deserialize<T>(buffer);
+        return _next.Deserialize(buffer);
     }
 }
 
+public sealed class NatsJsonContextSerializerRegistry : INatsSerializerRegistry
+{
+    private readonly JsonSerializerContext[] _contexts;
+
+    public NatsJsonContextSerializerRegistry(params JsonSerializerContext[] contexts) => _contexts = contexts;
+
+    public INatsSerializer<T> GetSerializer<T>() => new NatsJsonContextSerializer<T>(_contexts);
+}
+
 /// <summary>
 /// Serializer with support for <see cref="JsonSerializerContext"/>.
 /// </summary>
-public sealed class NatsJsonContextSerializer : INatsSerializer
+public sealed class NatsJsonContextSerializer<T> : INatsSerializer<T>
 {
     private static readonly JsonWriterOptions JsonWriterOpts = new() { Indented = false, SkipValidation = true };
 
     [ThreadStatic]
     private static Utf8JsonWriter? _jsonWriter;
 
-    private readonly JsonSerializerContext _context;
-    private readonly INatsSerializer? _next;
+    private readonly JsonSerializerContext[] _contexts;
+    private readonly INatsSerializer<T>? _next;
 
     /// <summary>
-    /// Creates a new instance of <see cref="NatsJsonContextSerializer"/>.
+    /// Creates a new instance of <see cref="NatsJsonContextSerializer{T}"/>.
     /// </summary>
-    /// <param name="context">Context to use for serialization.</param>
+    /// <param name="contexts">Context to use for serialization.</param>
     /// <param name="next">Next serializer in chain.</param>
-    public NatsJsonContextSerializer(JsonSerializerContext context, INatsSerializer? next = default)
+    public NatsJsonContextSerializer(JsonSerializerContext[] contexts, INatsSerializer<T>? next = default)
     {
-        _context = context;
+        _contexts = contexts;
         _next = next;
     }
 
+    public NatsJsonContextSerializer(JsonSerializerContext context, INatsSerializer<T>? next = default)
+        : this(new[] { context }, next)
+    {
+    }
+
     /// <inheritdoc />
-    public void Serialize<T>(IBufferWriter<byte> bufferWriter, T value)
+    public void Serialize(IBufferWriter<byte> bufferWriter, T value)
     {
-        if (_context.GetTypeInfo(typeof(T)) is JsonTypeInfo<T> jsonTypeInfo)
+        foreach (var context in _contexts)
         {
-            Utf8JsonWriter writer;
-            if (_jsonWriter == null)
+            if (context.GetTypeInfo(typeof(T)) is JsonTypeInfo<T> jsonTypeInfo)
```

Comment:
question: Is this always working as expected/intended? Could `GetTypeInfo` return a slightly different type than expected, i.e. an abstract type when `T` is a concrete type?

Reply:
I think so. Generated code is using `operator ==` on the type e.g.
```csharp
if (type == typeof(global::NATS.Client.Core.Internal.ServerInfo))
{
    return this.ServerInfo;
}
```